draft-irtf-cfrg-vdaf.md

```diff
@@ -3062,6 +3062,129 @@ measurement is sharded. This is provided to the FLP by Prio3.
+            weight = SumVec.Field(1 << j)
+            truncated[i] += weight * meas[i * SumVec.bits + j]
+    return truncated
+
+def decode(SumVec, output, _num_measurements):
+    return [x.as_unsigned() for x in output]
+~~~
+
+This validity circuit uses a `ParallelSum` gadget to achieve a smaller proof
+size. This optimization for "parallel-sum circuits" is described in
+{{BBCGGI19}}, section 4.4. Briefly, for circuits that add up the output of
+multiple identical sub-circuits, it is possible to achieve smaller proof sizes
+(on the order of O(sqrt(n)) instead of O(n)) by packaging more than one such
+subcircuit into a gadget.
+
+The `ParallelSum` gadget is parameterized with an arithmetic subcircuit, and a
+`count` of how many times it evaluates that subcircuit. It takes in a list of
+inputs and passes them through to instances of the subcircuit in the same order.
+It returns the sum of the subcircuit outputs.
+
+~~~
+def ParallelSum(self, inputs):
+    out = Field(0)
+    for i in range(self.count):
+        start_index = i * self.subcircuit.ARITY
+        end_index = (i + 1) * self.subcircuit.ARITY
+        out += self.subcircuit(inputs[start_index:end_index])
+    return out
+~~~
+
+The validity circuit checks that the encoded measurement consists of ones and
+zeros. Rather than use the `Range2` gadget on each element, it instead uses a
+`Mul` subcircuit and "free" constant multiplication and addition gates to
+simultaneously evaluate the same range check polynomial on each element, and
+multiply by a constant. (Arranging the circuit this way allows us to use the
+`ParallelSum` gadget, because the summands are identical subcircuits. Using
+`Range2` and then multiplying by different constants would not fit the parallel
+sum pattern.) One of the two `Mul` subcircuit inputs is equal to a measurement
+element multiplied by a power of the joint randomness value, and the other is
+equal to the same measurement element minus one. These `Mul` subcircuits are
+evaluated by a `ParallelSum` gadget, and the results are added up both within
+the `ParallelSum` gadget and after it.
+
+~~~
+def SumVec(self,
+           meas: Vec[Field128],
+           joint_rand: Vec[Field128],
+           num_shares: Unsigned):
+    out = Field128(0)
+    r = joint_rand[0]
+    r_power = r
+    shares_inv = Field128(num_shares).inv()
+
+    gadget_calls = (
+        (self.length * self.bits + self.chunk_length - 1)
+        // self.chunk_length
+    )
+    for i in range(gadget_calls):
+        inputs = [None] * (2 * self.chunk_length)
+        for j in range(self.chunk_length):
+            index = i * self.chunk_length + j
+            if index < len(meas):
+                meas_elem = meas[index]
+            else:
+                meas_elem = self.Field(0)
+
+            inputs[j * 2] = r_power * meas_elem
+            inputs[j * 2 + 1] = meas_elem - shares_inv
+
+            r_power *= r
+
+        out += ParallelSum(Mul(), self.chunk_length, inputs)
```

Comment:
* How is `ParallelSum.count` computed from `self.chunk_length`? It looks to me like what we mean to is `len(inputs) / self.chunk_length` (`+1` if there is a partial chunk)
* Hmm, we have a class method called `ParrallelSum` on an undefined class and a constructor for a class called `ParallelSum`. This seems to suggest an odd edge case in how we've been specifying circuits up to this point. This suggests to me that we may want to refactor the API a bit. What do you think about renaming the former to a class method called `call()`?

Reply:
The chunk_length VDAF parameter is passed directly through to be `ParallelSum.count`. The gadget's `count` parameter is the number of times it calls the circuit, see the description above. We do that division by chunk_length to decide how many times to call the ParallelSum gadget instead.

Yeah, all the pseudocode for gadgets or validity circuits thus far doesn't match the Python implementation, as they're written as if they are very short functions. The shorthand is breaking down with ParallelSum, since it takes parameters upon its construction, and takes a variable number of input wires. I think making the distinction between the class and the `call()` method under the covers would clear this up.

Reply:
Do you want to take a stab at adding `call()` this PR, or do you think we should do it in a separate PR and rebase? Your call -- I think it's a pretty minor change and impcts the reference code more than the draft,

Reply:
Yeah, a separate PR sounds like a good idea, I keep finding things I want to clean up. (like fixing shadowing class names with the first parameter to class methods, and moving check_valid_eval etc. to be superclass methods instead of free functions)

Reply:
Rebased over #280